app/models/ext_management_system.rb

```diff
@@ -109,6 +112,22 @@ def hostname_format_valid?
     errors.add(:hostname, _("format is invalid."))
   end
 
+  # validation - Zone cannot be changed when enabled == false
+  def validate_ems_enabled_when_zone_changed?
+    return if changed_attributes.include?('enabled')
+
+    if changed_attributes.include?('zone_id') && !enabled?
+      errors.add(:zone, N_("cannot be changed when provider paused"))
+    end
+  end
+
+  # validation - Zone has to be visible when enabled == true
+  def validate_zone_visible_when_ems_enabled?
+    if enabled? && zone.present? && !zone.visible?
+      errors.add(:zone, N_("cannot be invisible when provider active"))
```

Comment:
It seems like you're conflating the zone being invisible and the zone being the maintenance zone.  Would this make more sense saying `enabled? && zone.maintenance?`

Reply:
@agrare actually maintenance zone is only member of invisible zones. This comcept brings invisible zones for various purposes, one of them is hidden maintenance zone. But I can imagine another invisible zone for some special operation

Reply:
I agree that it is possible for us to use an invisible zone in the future, which is exactly why i think we shouldn't say you can't add an enabled provider to an invisible zone.

We might want to be able to add a provider to an internal/invisible zone.

Reply:
While we'd like to make visibility a future thing, we don't have use cases in the linked BZs.  Perhaps there's another BZ RFE for visibility but if not, I think we should concentrate on working on the only thing requested: ability to put our providers into a maintenance zone so we do not try to connect to them via pause/resume.

I'm having great difficulty understanding what is desired in this PR because visibility is being checked when what we really care about is if the provider is active or paused.

Reply:
To build on that, the UI should be making decisions about visibility based on the state of the provider... if it's paused/active/whatever we call it... I'm super late to the game but I don't see how visible is a backend/column concern.

Reply:
@jrafanie UI makes decisions about visibility not only by state of provider, it needs to know about it in select inputs without associated EMS. 1) Creating new provider 2) List of zones in Configuration. 3) Assigning server to zone in configuration's select.

For UI, it's better to use some generic flag than naming and excluding "maintenance"

Provider's state in UI is checked primarily by `ExtManagementSystem.enabled` (because children EMS's are not in maintenance zone - was it your request, @agrare?)

Reply:
The linked BZ has 18 references to maintenance and 0 references to visible or visibility.

What the UI wants to do with regard to hiding/showing objects should be based on the only requirement in the BZ, maintenance mode/paused.

This is why it's hard for me to understand some of the validations and code decisions in the PR, because I have no idea what visible means.

Reply:
> because children EMS's are not in maintenance zone - was it your request, @agrare?

Pretty sure not, child EMSs not being in maintenance is a problem because they will continue to service requests.  That's why we were suggesting calling `pause!` not `disable!` on the child emss down [here](https://github.com/ManageIQ/manageiq/pull/17452/files#r227077922)

Reply:
@jrafanie while the BZ may not state visibility, it's an obvious corollary.  If you have a zone that will ultimately present itself in dropdowns, but that zone should not be modified manually, then there needs to be a way to remove it from dropdowns.  It's an inherent feature of the maintenance zone, so it can't really be a separate PR.

Reply:
This PR should deal with maintenance mode (or enabled) and on top of that, visibility can be added by checking for it.  I don't understand how these validations correspond with the pause/resume actions in the UI.  You pause/resume the state/enabled, not visibility.  Visibility is a side effect of being enabled or not, right?

Reply:
See also: https://github.com/ManageIQ/manageiq/pull/17452#issuecomment-432356848